autoload/arduino.vim

```diff
@@ -118,19 +121,36 @@ endfunction
 function! arduino#GetBoards()
   let arduino_dir = arduino#GetArduinoDir()
   let boards = []
-  for filename in split(globpath(arduino_dir . '/hardware', '**/boards.txt'), '\n')
-    let pieces = split(filename, '/')
-    let package = pieces[-3]
-    let arch = pieces[-2]
-    let lines = readfile(filename)
-    for line in lines
-      if line =~? '^[^.]*\.build\.board=.*$'
-        let linesplit = split(line, '\.')
-        let board = linesplit[0]
-        call add(boards, package . ':' . arch . ':' . board)
-      endif
+  if arduino#GetArduinoUserInstallation() == 1
+    for filename in split(globpath(arduino_dir . '/packages', '**/boards.txt'), '\n')
+      let pieces = split(filename, '/')
+      let package = pieces[-5]
+      let arch = pieces[-3]
+      let package_version = pieces[-2]
+      let lines = readfile(filename)
+      for line in lines
+        if line =~? '^[^.]*\.build\.board=.*$'
+          let linesplit = split(line, '\.')
+          let board = linesplit[0]
+          call add(boards, package . ':' . arch . ':' . board . ':' . package_version)
+        endif
+      endfor
     endfor
-  endfor
+  else
+    for filename in split(globpath(arduino_dir . '/hardware', '**/boards.txt'), '\n')
+      let pieces = split(filename, '/')
+      let package = pieces[-3]
+      let arch = pieces[-2]
+      let lines = readfile(filename)
+      for line in lines
+        if line =~? '^[^.]*\.build\.board=.*$'
+          let linesplit = split(line, '\.')
+          let board = linesplit[0]
+          call add(boards, package . ':' . arch . ':' . board)
+        endif
+      endfor
+    endfor
+  endif
```

Comment:
Could you take the shared logic here and factor it out into a function?

Reply:
Done